core/vcpu.c

```diff
@@ -2575,10 +2583,31 @@ static void handle_cpuid_virtual(struct vcpu_t *vcpu, uint32_t a, uint32_t c)
              * that is an old i7 system, so the emulator can still utilize the
              * enough extended features of the hardware, but doesn't crash.
              */
-            hw_family = ((state->_eax >> 16) & 0xFF0) |
-                        ((state->_eax >> 8) & 0xF);
-            hw_model = ((state->_eax >> 12) & 0xF0) |
-                       ((state->_eax >> 4) & 0xF);
+             union cpuid_1_eax {
+                 uint32_t raw;
+                 struct {
+                     unsigned steppingID    : 4;
+                     unsigned model         : 4;
+                     unsigned familyID      : 4;
+                     unsigned processorType : 2;
+                     unsigned reserved      : 2;
+                     unsigned extModelID    : 4;
+                     unsigned extFamilyID   : 8;
+                     unsigned reserved2     : 4;
+                 };
+             } cpuid_eax;
+             cpuid_eax.raw = state->_eax;
```

Comment:
Have some alignment issue here.

Reply:
On which platforms? I think it would be better to change unsigned to uint32_t.

Reply:
Yes, to align with the type of raw could eliminate the reading ambiguity.
And I didn't mean byte alignment here but the format. :smile: There seems an extra space before the piece of code.